### PR TITLE
Drop TS dependencies from cummeRbund tool, use

### PR DESCRIPTION
**Pull request by mvdbeek**
`<command detect_errors="aggressive">`

### CODE REVIEW

File: tools/cummerbund/cummeRbund.xml [hunk 1]
```diff
@@ -1,20 +1,17 @@
 <?xml version="1.0"?>
-<tool id="cummeRbund" name="cummeRbund" version="1.0.2">
+<tool id="cummeRbund" name="cummeRbund" version="2.16.2">
     <description>visualize Cuffdiff output</description>
     <requirements>
-        <requirement type="set_environment">CUMMERBUND_SCRIPT_PATH</requirement>
-        <requirement type="package" version="3.1.2">R</requirement>
-        <requirement type="package" version="2.8.2">cummeRbund</requirement>
+        <requirement type="package" version="2.16.2">cummeRbund</requirement>
```
Comment (lecorguille) [inline tools/cummerbund/cummeRbund.xml]: Tu m'as doublé :P 
```xml
        <requirement type="package" version="1.0.1">r-argparse</requirement>
        <requirement type="package" version="2.8.2">bioconductor-cummerbund</requirement>
```

File: tools/cummerbund/cummeRbund.xml [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
-    <expand macro="stdio" />
     <macros>
         <import>cummeRbund_macros.xml</import>
     </macros>
     <code file="cummeRbund_options.py"/>
-    <command>
+    <command detect_errors="aggressive">
 <![CDATA[
 #for i, p in enumerate($plots):
-    R --vanilla --no-save -f \$CUMMERBUND_SCRIPT_PATH/cummeRbund.R --args
```
Comment (lecorguille) [inline tools/cummerbund/cummeRbund.xml]: You also need to drop `--args`

**Comment by KeywanHP** [comment]
Great thanks.

File: tools/cummerbund/.shed.yml [hunk 3]
```diff
@@ -8,5 +8,6 @@ long_description: |
   functions for commonly used visualizations.
 name: cummerbund
 owner: devteam
+homepage_url: http://compbio.mit.edu/cummeRbund
```
Comment (nsoranzo) [inline tools/cummerbund/.shed.yml]: What about https://bioconductor.org/packages/release/bioc/html/cummeRbund.html instead?

Comment (mvdbeek) [inline tools/cummerbund/.shed.yml]: seems more, recent i'll add that

**Comment by lecorguille** [comment]
I added 
`options(echo=TRUE)` at the beginning of the cummerRbund.R to get log as close as they were.
You also need to update the .png but `planemo --update_test_data` doen't want to update the `<discovered_dataset`
And there is still an issue or me with `heatmap`
```
Using tracking_id, sample_name as id variables
Failed: Error in mat[, i]: subscript out of bounds
```

**Comment by mvdbeek** [comment]
Yeah I think this is the last pice of the puzzle. I have no clue what's wrong there.

**Comment by mvdbeek** [comment]
OK, it's refusing to draw a heatmap for a single gene. Rightly so. 👏 . The previous version was just drawing a heatmap for all genes.

**Comment by lecorguille** [comment]
Well done

**Comment by KeywanHP** [comment]
👍 Is it now available in main TS?

**Comment by nsoranzo** [comment]
It will be when this Travis build will finish: https://travis-ci.org/galaxyproject/tools-devteam/builds/204697063

**Comment by KeywanHP** [comment]
Installed this new version. But for some reason the output is R code rather than plots.

```
Rscript /home/usern/galaxy/shed_tools/toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/78fcfc04fcfe/cummerbund/cummeRbund.R --input "/home/usern/galaxy/galaxy/database/files/000/139/dataset_139264.dat" --width 1280 --height 960 --type "maplot" --outfile plot-maplot-0.png --x "q1" --y "q2" > "/home/usern/galaxy/galaxy/database/files/000/139/dataset_139265.dat" 2>&1 ;
```

```
> png(filename = args$filename, width = args$width, height = args$height, type=c('cairo-png'))
Warning messages:
1: In png(filename = args$filename, width = args$width, height = args$height,  :
  unable to load shared object '/home/usern/galaxy/galaxy/database/dependencies/_conda/envs/mulled-v1-71206c22effe4f02960f526961e99857b4140cdce3da5f9a9e75417fc1a3e405/lib/R/library/grDevices/libs//cairo.so':
  libXrender.so.1: cannot open shared object file: No such file or directory
2: In png(filename = args$filename, width = args$width, height = args$height,  :
  failed to load cairo DLL
```

**Comment by mvdbeek** [comment]
The png output would be dynamically discovered. The problem is that writing the png output failed because cairo was attempting to use some function from libXrender. I suppose that didn't happend with the toolshed R version because we disabled X11 explicitly. I think I have seen a solution for this a while ago ...

**Comment by mvdbeek** [comment]
Could you have a look at the conda channel order that your galaxy is using?
According to https://github.com/ContinuumIO/anaconda-issues/issues/1063 there may be an issue if the channel order is not correct. The correct order is specified [here](https://github.com/galaxyproject/galaxy/blob/dev/config/galaxy.ini.sample#L212). This list should also be written out to the _condarc file (which should be in your tool dependency folder, by default in database/dependencies). That said I'm not sure that any X11 libraries installed on travis may have been "bleeding" through into planemo's test environment, giving us a false positive. @bgruening do you remember a trick to turn off cairo's X11 capabilities?

**Comment by KeywanHP** [comment]
Done what you have suggested but the problem still exists...

conda_ensure_channels = iuc,bioconda,r,defaults,conda-forge

**Comment by lecorguille** [comment]
Are you sure that Galaxy is using the conda package rather the system R installation?
Can you give us the launch line from your logs?

**Comment by KeywanHP** [comment]
```
galaxy.tools DEBUG 2017-02-27 11:11:01,972 Validated and populated state for tool request (420.235 ms)
galaxy.tools.actions INFO 2017-02-27 11:11:02,097 Handled output named output for tool toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0 (51.333 ms)
galaxy.tools.actions INFO 2017-02-27 11:11:02,108 Added output datasets to history (9.458 ms)
galaxy.tools.actions INFO 2017-02-27 11:11:02,145 Verified access to datasets for Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0] (18.879 ms)
galaxy.tools.actions INFO 2017-02-27 11:11:02,148 Setup for job Job[unflushed,tool_id=toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0] complete, ready to flush (38.498 ms)
galaxy.tools.actions INFO 2017-02-27 11:11:02,200 Flushed transaction for job Job[id=59478,tool_id=toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0] (50.893 ms)
galaxy.tools.execute DEBUG 2017-02-27 11:11:02,201 Tool [toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0] created job [59478] (183.441 ms)
galaxy.tools.execute DEBUG 2017-02-27 11:11:02,218 Executed 1 job(s) for tool toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0 request: (244.753 ms)
galaxy.tools.parameters.basic DEBUG 2017-02-27 11:11:07,388 Error determining dynamic options for parameter 'x' in tool 'toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0':
Traceback (most recent call last):
  File "/home/usern/galaxy/galaxy/lib/galaxy/tools/parameters/basic.py", line 787, in get_options
    return eval( self.dynamic_options, self.tool.code_namespace, call_other_values )
  File "<string>", line 1, in <module>
NameError: name 'input_database' is not defined
galaxy.tools.parameters.basic DEBUG 2017-02-27 11:11:07,393 Error determining dynamic options for parameter 'y' in tool 'toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0':
Traceback (most recent call last):
  File "/home/usern/galaxy/galaxy/lib/galaxy/tools/parameters/basic.py", line 787, in get_options
    return eval( self.dynamic_options, self.tool.code_namespace, call_other_values )
  File "<string>", line 1, in <module>
NameError: name 'input_database' is not defined
galaxy.tools.parameters.basic DEBUG 2017-02-27 11:11:07,399 Error determining dynamic options for parameter 'gene_id' in tool 'toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0':
Traceback (most recent call last):
  File "/home/usern/galaxy/galaxy/lib/galaxy/tools/parameters/basic.py", line 787, in get_options
    return eval( self.dynamic_options, self.tool.code_namespace, call_other_values )
  File "<string>", line 1, in <module>
NameError: name 'input_database' is not defined
galaxy.tools.parameters.basic DEBUG 2017-02-27 11:11:07,404 Error determining dynamic options for parameter 'x' in tool 'toolshed.g2.bx.psu.edu/repos/devteam/cummerbund/cummeRbund/2.16.0':
```

**Comment by KeywanHP** [comment]
It doesn't really show if Galaxy is using conda packages or the system installation of R... Is there a way to find that out?

**Comment by mvdbeek** [comment]
This is quite a different error form what you posted above; in this current call you're not even executing the tool because something is wrong with the tool parameters.

**Comment by KeywanHP** [comment]
I'm not sure what it does.. The tool goes yellow and after a while produces the attached output...

[Galaxy126-[cummeRbund_on_data_120].txt](https://github.com/galaxyproject/tools-devteam/files/803460/Galaxy126-.cummeRbund_on_data_120.txt)

I wonder if something is wrong with our Galaxy setup?

**Comment by lecorguille** [comment]
Can you give us the whole log from the click on execute to final (green or red) dataset state when the job is done?
You can use https://gist.github.com/ if it's huge.

**Comment by mvdbeek** [comment]
Look for something like `[galaxy.tools.deps] Using dependency bioconductor-cummerbund version 2.16.0 of type conda` in the logs. And try using this file as input database `https://github.com/galaxyproject/tools-devteam/blob/master/tools/cummerbund/test-data/cuffdiff_out.sqlite`

**Comment by KeywanHP** [comment]
Here is the gist of handler and main log
https://gist.github.com/KeywanHP/019494fab7f7aaf336a72fafb99decc4.js

**Comment by mvdbeek** [comment]
This is not the complete log output, though I can see that your environment is picking up mulled-v1-71206c22effe4f02960f526961e99857b4140cdce3da5f9a9e75417fc1a3e405, so I would presume that's the environment for cummerbund. After changing the channel order did you delete the environment and re-install the tool? Please keep a log of this as well.

**Comment by KeywanHP** [comment]
All I filtered out are the lines containing Gecko. Do you want to see those as well?
tail -f /home/usern/galaxy/galaxy/main.log | grep -v Gecko

**Comment by lecorguille** [comment]
Can you try that:
```
source /home/usern/galaxy/galaxy/
database/dependencies/_conda/bin/activate mulled-v1-71206c22effe4f02960f526961e99857b4140cdce3da5f9a9e75417fc1a3e405
R
```
```R
> library(cummeRbund)
> capabilities()
```

**Comment by KeywanHP** [comment]
```
> capabilities()
       jpeg         png        tiff       tcltk         X11        aqua
       TRUE        TRUE        TRUE        TRUE       FALSE       FALSE
   http/ftp     sockets      libxml        fifo      cledit       iconv
       TRUE        TRUE        TRUE        TRUE        TRUE        TRUE
        NLS     profmem       cairo         ICU long.double     libcurl
       TRUE        TRUE        TRUE        TRUE        TRUE        TRUE
Warning message:
In doTryCatch(return(expr), name, parentenv, handler) :
  unable to load shared object '/home/usern/galaxy/galaxy/database/dependencies/_conda/envs/mulled-v1-71206c22effe4f02960f526961e99857b4140cdce3da5f9a9e75417fc1a3e405/lib/R/modules//R_X11.so':
  libXt.so.6: cannot open shared object file: No such file or directory
```

**Comment by bgruening** [comment]
@KeywanHP looks like a conda package problem. I guess we have not defined the X libraries as dependencies - or we should not use the X backend in this wrapper.

**Comment by KeywanHP** [comment]
I guess there are other wrappers that use the X backend and would fail on our system. Probably the better solution would be to add X libraries as dependencies?

**Comment by mvdbeek** [comment]
I think there is a way to tell the cairo backend not to use X11 at runtime, I just can't find this anymore. The problem with including X11 dependencies is that not everyone wants or needs them. 

About the particular problem here: This is a warning message in a `doTryCatch` clause, it may not actually be fatal. Could you change `<command detect_errors="aggressive">` to `<command detect_errors="exit_code">` and see if you get the correct output then?

@bgruening @jmchilton so I think we got our first case of a false positive test because of the things that are preinstalled on travis. Testing in containers would be the obvious answer here. Do you agree with that?

**Comment by mvdbeek** [comment]
So the container has a similar problem, which makes the debugging easier.
```
docker run -it quay.io/biocontainers/bioconductor-cummerbund:2.16.0--r3.3.1_0 sh
/ # R

R version 3.3.1 (2016-06-21) -- "Bug in Your Hair"
Copyright (C) 2016 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> capabilities()
       jpeg         png        tiff       tcltk         X11        aqua
       TRUE        TRUE        TRUE        TRUE       FALSE       FALSE
   http/ftp     sockets      libxml        fifo      cledit       iconv
       TRUE        TRUE        TRUE        TRUE        TRUE        TRUE
        NLS     profmem       cairo         ICU long.double     libcurl
       TRUE        TRUE        TRUE        TRUE        TRUE        TRUE
Warning message:
In doTryCatch(return(expr), name, parentenv, handler) :
  unable to load shared object '/usr/local/lib/R/modules//R_X11.so':
  libSM.so.6: cannot open shared object file: No such file or directory
>
```

**Comment by mvdbeek** [comment]
@KeywanHP you don't need to try changing `<command detect_errors="aggressive">` to `<command detect_errors="exit_code">` in cummerbund.xml, it won't work. I'm testing if rebuilding bioconductor-cummerbund is enough to fix this.

**Comment by mvdbeek** [comment]
Args, so this is simple broken in r-base, affected is also the latest conda-forge `r-base:     3.3.2-4       conda-forge` r, so no solution in sight. Two workarounds are possible, we can switch to PDF output, or we add a dependency on ghostscript and use `bitmap(filename,"png16m")` as mentioned by @yhoogstrate on biostars https://biostar.usegalaxy.org/p/9170/. Any preference? I think using PDF output is not too bad.

**Pull request merged**